File: cafe-recommend/backend/app/routers/user_identity.py
from fastapi import APIRouter, Depends, Request, Response, HTTPException
from pydantic import BaseModel
from typing import Optional, Dict, Any
import uuid

from app.core.session import SessionData, get_session, session_manager

router = APIRouter(
    prefix="/user",
    tags=["user-identity"],
    responses={404: {"description": "Not found"}},
)

class UserIdentityResponse(BaseModel):
    user_id: str
    is_new: bool

class UserPreferences(BaseModel):
    theme: Optional[str] = None
    language: Optional[str] = None
    notification_enabled: Optional[bool] = None

@router.get("/identify", response_model=UserIdentityResponse)
async def identify_user(
    request: Request,
    response: Response,
    session: SessionData = Depends(get_session)
):
    """
    사용자 식별자 가져오기 또는 생성하기
    """
    is_new_user = False
    user_id = session.get("user_id")
    
    if not user_id:
        # 새 사용자 ID 생성
        user_id = str(uuid.uuid4())
        session["user_id"] = user_id
        is_new_user = True
    
    # 세션 저장
    session_manager.save_session(response, session)
    
    return UserIdentityResponse(
        user_id=user_id,
        is_new=is_new_user
    )

@router.get("/preferences", response_model=Dict[str, Any])
async def get_user_preferences(
    request: Request,
    session: SessionData = Depends(get_session)
):
    """
    사용자의 저장된 설정 가져오기
    """
    user_id = session.get("user_id")
    if not user_id:
        # 사용자 ID가 없는 경우에도 기본 설정 반환
        return {
            "theme": "light",
            "language": "ko",
            "notification_enabled": True
        }
    
    preferences = session.get("preferences", {})
    # 기본값이 없는 경우 기본 설정 추가
    if not preferences:
        preferences = {
            "theme": "light",
            "language": "ko",
            "notification_enabled": True
        }
    return preferences

@router.post("/preferences")
async def update_user_preferences(
    preferences: UserPreferences,
    request: Request,
    response: Response,
    session: SessionData = Depends(get_session)
):
    """
    사용자 설정 업데이트
    """
    user_id = session.get("user_id")
    if not user_id:
        raise HTTPException(status_code=404, detail="사용자를 찾을 수 없습니다")
    
    # 기존 설정 가져오기
    current_prefs = session.get("preferences", {})
    
    # 새 설정으로 업데이트
    updated_prefs = {**current_prefs}
    
    if preferences.theme is not None:
        updated_prefs["theme"] = preferences.theme
    
    if preferences.language is not None:
        updated_prefs["language"] = preferences.language
    
    if preferences.notification_enabled is not None:
        updated_prefs["notification_enabled"] = preferences.notification_enabled
    
    # 세션에 저장
    session["preferences"] = updated_prefs
    session_manager.save_session(response, session)
    
    return {"status": "success", "preferences": updated_prefs}

@router.post("/clear")
async def clear_user_session(
    request: Request,
    response: Response
):
    """
    사용자 세션 초기화
    """
    session_manager.clear_session(response)
    return {"status": "success", "message": "사용자 세션이 초기화되었습니다"} 

File: cafe-recommend/frontend/src/lib/user-identity.ts
/**
 * 사용자 식별 시스템 관련 API 클라이언트
 */

import apiClient from './api-client';

// 사용자 식별 응답 타입
export interface UserIdentityResponse {
  user_id: string;
  is_new: boolean;
}

// 사용자 설정 타입
export interface UserPreferences {
  theme?: string;
  language?: string;
  notification_enabled?: boolean;
}

/**
 * 사용자 ID를 가져오거나 생성합니다.
 * 처음 방문한 사용자는 새 ID가 생성되고, 재방문 사용자는 기존 ID가 반환됩니다.
 */
export async function identifyUser(): Promise<UserIdentityResponse> {
  try {
    const response = await apiClient.get<UserIdentityResponse>('/api/user/identify');
    return response;
  } catch (error) {
    console.error('사용자 식별 실패:', error);
    throw error;
  }
}

/**
 * 사용자 설정을 가져옵니다.
 */
export async function getUserPreferences(): Promise<UserPreferences> {
  try {
    const response = await apiClient.get<UserPreferences>('/api/user/preferences');
    return response;
  } catch (error) {
    console.error('사용자 설정 가져오기 실패:', error);
    // 기본 설정 반환
    return {
      theme: 'light',
      language: 'ko',
      notification_enabled: true
    };
  }
}

/**
 * 사용자 설정을 업데이트합니다.
 */
export async function updateUserPreferences(preferences: UserPreferences): Promise<UserPreferences> {
  try {
    const response = await apiClient.post<{status: string, preferences: UserPreferences}>('/api/user/preferences', {
      data: preferences
    });
    return response.preferences;
  } catch (error) {
    console.error('사용자 설정 업데이트 실패:', error);
    throw error;
  }
}

/**
 * 사용자 세션을 초기화합니다.
 */
export async function clearUserSession(): Promise<{status: string, message: string}> {
  try {
    const response = await apiClient.post<{status: string, message: string}>('/api/user/clear');
    return response;
  } catch (error) {
    console.error('사용자 세션 초기화 실패:', error);
    throw error;
  }
} 